src/main/scala/org/apache/spark/sql/pulsar/PulsarSourceRDD.scala

```diff
@@ -124,12 +134,63 @@ private[pulsar] abstract class PulsarSourceRDDBase(
           throw e
       }
 
+      private def processDataLoss(
+          currentMessageId: MessageId,
+          previousMessageId: MessageId): Unit = {
+
+        reportDataLoss(
+          s"Unexpected message skipping was detected:" +
+            s" Previous message id was $previousMessageId," +
+            s" however the current message read has id $currentMessageId"
+        )
+      }
+
+      /**
+       * Detect data loss by comparing the current message id with the previous message id.
+       * Make sure invariants are held
+       */
+      private def detectDataLoss(
+          currentMessageId: MessageId,
+          previousMessageId: MessageId): Unit = {
+        (currentMessageId, previousMessageId) match {
+          case (c: BatchMessageIdImpl, p: BatchMessageIdImpl) =>
+            if (c.getLedgerId == p.getLedgerId) {
+              if (c.getEntryId == p.getEntryId) {
+                if (c.getBatchIndex != p.getBatchIndex + 1) {
+                  processDataLoss(c, p)
+                }
+              } else if (c.getEntryId == p.getEntryId + 1) {
+                // if we have moved to ready the next entry
+                // the batch index should be 0, i.e. to first record
+                // in the entry / batch
+                if (c.getBatchIndex != 0) {
+                  processDataLoss(c, p)
+                }
+              } else if (c.getEntryId != p.getEntryId + 1) {
+                processDataLoss(c, p)
+              }
+            }
+          case (c: MessageIdImpl, p: BatchMessageIdImpl) =>
```

Comment:
You missed two cases:

```scala
case (c: BatchMessageIdImpl, p: MessageIdImpl) =>
  // If ledgerIds are the same, confirm current id's entry id should be "previous +1"
case (c: MessageIdImpl, p: BatchMessageIdImpl) =>
  //The variable `batchIndex` of the latest single message of a batch should be `batchSize -1`
```

BTW, I consider that the following logic is clearer: 
- If `ledgerId and entryId` are the same, compare `batch index`
- otherwise:
  - `ledgerId` is the same: compare entry id and confirm that the `batch index` is `0`
  - `ledgerId` is not the same: skip to compare

---

Important knowledge: if users send an empty message, it will be discarded by the client library, which means that you assumed it is a message lost, but the client loses nothing.

Reply:
added checks for those scenarios

Reply:
What is the scenario you are referring to?  How can an empty message be sent?